apps/rush-lib/src/logic/pnpm/PnpmShrinkwrapFile.ts

```diff
@@ -182,22 +196,84 @@ export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
     }
   }
 
-  public static loadFromFile(shrinkwrapYamlFilename: string): PnpmShrinkwrapFile | undefined {
+  public static loadFromFile(
+    shrinkwrapYamlFilename: string,
+    pnpmOptions: PnpmOptionsConfiguration
+  ): PnpmShrinkwrapFile | undefined {
     try {
       if (!FileSystem.exists(shrinkwrapYamlFilename)) {
         return undefined; // file does not exist
       }
 
       // We don't use JsonFile/jju here because shrinkwrap.json is a special NPM file format
       // and typically very large, so we want to load it the same way that NPM does.
-      const parsedData: IPnpmShrinkwrapYaml = yaml.safeLoad(FileSystem.readFile(shrinkwrapYamlFilename).toString());
+      const shrinkwrapContent: string = FileSystem.readFile(shrinkwrapYamlFilename).toString();
+      const parsedData: IPnpmShrinkwrapYaml = yaml.safeLoad(shrinkwrapContent);
+
+      let shrinkwrapHash: string | undefined;
+      if (pnpmOptions.preventManualShrinkwrapChanges) {
+        // Grab the shrinkwrap hash out of the comment where we store it.
+        const hashYamlCommentRegExp: RegExp = new RegExp(
+          `\\n\\s*${PnpmShrinkwrapFile._shrinkwrapHashPrefix}\\s*(\\S+)\\s*$`
+        );
+        const match: RegExpMatchArray | null = shrinkwrapContent.match(hashYamlCommentRegExp);
+        shrinkwrapHash = match ? match[1] : undefined;
+      }
 
-      return new PnpmShrinkwrapFile(parsedData, shrinkwrapYamlFilename);
+      return new PnpmShrinkwrapFile(
+        parsedData,
+        shrinkwrapYamlFilename,
+        shrinkwrapHash,
+        pnpmOptions.preventManualShrinkwrapChanges
+      );
     } catch (error) {
       throw new Error(`Error reading "${shrinkwrapYamlFilename}":${os.EOL}  ${error.message}`);
     }
   }
 
+  /** @override */
+  public shouldForceRecheck(): boolean {
+    // Ensure the shrinkwrap is rechecked when the hash is enabled but no hash is populated.
+    return super.shouldForceRecheck() || (!!this._shrinkwrapHashEnabled && !this._shrinkwrapHash);
+  }
+
+  /** @override */
+  public validate(
+    packageManagerOptionsConfig: PackageManagerOptionsConfigurationBase,
+    policyOptions: IPolicyValidatorOptions
+  ) : void {
+    super.validate(packageManagerOptionsConfig, policyOptions);
+    if (!(packageManagerOptionsConfig instanceof PnpmOptionsConfiguration)) {
+      return;
+    }
+
+    // Only check the hash if allowShrinkwrapUpdates is false. If true, the shrinkwrap file
+    // may have changed and the hash could be invalid.
+    if (packageManagerOptionsConfig.preventManualShrinkwrapChanges && !policyOptions.allowShrinkwrapUpdates) {
+      if (!this._shrinkwrapHash) {
+        console.log(
+          colors.red(
+            'The shrinkwrap file does not contain the generated hash. You may need to run "rush update" to ' +
+            'populate the hash.'
```

Comment:
I kind of liked the idea of mentioning `preventManualShrinkwrapChanges` to help people find help for this error, something like:
```ts
            'The shrinkwrap file does not contain the generated hash. You may need to run "rush update" to ' +
            'populate the hash. See the "preventManualShrinkwrapChanges" setting documentation for details.'
```